src/sqlancer/postgres/ast/PostgresJoin.java

```diff
@@ -15,18 +17,36 @@ public static PostgresJoinType getRandom() {
 
     }
 
-    private final PostgresTable table;
+    private final PostgresTableReference tableReference;
     private final PostgresExpression onClause;
     private final PostgresJoinType type;
 
-    public PostgresJoin(PostgresTable table, PostgresExpression onClause, PostgresJoinType type) {
-        this.table = table;
+    public static class PostgresTableReference implements PostgresExpression {
+
+        private final PostgresExpression tableReference;
+
+        public PostgresTableReference(PostgresCTE cte) {
+            this.tableReference = cte;
+        }
+
+        public PostgresTableReference(PostgresTable table) {
+            this.tableReference = new PostgresFromTable(table, Randomly.getBoolean());
+        }
+
+        public PostgresExpression getTableReference() {
+            return tableReference;
+        }
+
+    }
+
+    public PostgresJoin(PostgresTableReference tableReference, PostgresExpression onClause, PostgresJoinType type) {
```

Comment:
Related to the comment above: I think we should use a `PostgresExpression` type here.